spikeinterface/sortingcomponents/matching/wobble.py

```diff
@@ -0,0 +1,948 @@
+        at_least_one_spike = bool(len(non_refractory_indices))
+        if at_least_one_spike:
+            spike_jittered_ids[non_refractory_indices] += template_shift
+            spike_time_indices[non_refractory_indices] += time_shift
+            scalings[non_refractory_indices] = scaling
+
+        # Generate new spike train from spike times (indices)
+        convolution_correction = -1*(template_meta.n_time - 1) # convolution indices --> raw_indices
+        spike_time_indices += convolution_correction
+        new_spike_train = np.array([spike_time_indices, spike_jittered_ids]).T
+
+        return new_spike_train, scalings, distance_metric
+
+
+    @classmethod
+    def subtract_spike_train(cls, spike_train, scalings, objective, params, template_meta, sparsity):
+        """Subtract spike train of templates from the objective directly.
+
+        Operates in-place on the objective.
+
+        Parameters
+        ----------
+        spike_train : ndarray (n_spikes, 2)
+            Spike train from template matching.
+        scalings : ndarray (n_spikes,)
+            Amplitude scaling used for each spike.
+        objective : Objective
+            Dataclass object for aggregating variables related to the objective together.
+        params : WobbleParameters
+            Dataclass object for aggregating the parameters together.
+        template_meta : TemplateMetadata
+            Dataclass object for aggregating template metadata together.
+        sparsity : Sparsity
+            Dataclass object for aggregating channel sparsity variables together.
+        """
+        present_jittered_ids = np.unique(spike_train[:, 1])
+        convolution_resolution_len = get_convolution_len(template_meta.n_time, template_meta.n_time)
+        for jittered_id in present_jittered_ids:
+            id_mask = spike_train[:, 1] == jittered_id
+            id_spiketrain = spike_train[id_mask, 0]
+            id_scaling = scalings[id_mask]
+            overlapping_templates = sparsity.unit_overlap[jittered_id]
+            # Note: pairwise_conv only has overlapping template convolutions already
+            pconv = objective.pairwise_convolution[jittered_id]
+            # TODO: If optimizing for speed -- check this loop
+            for spike_start_idx, spike_scaling in zip(id_spiketrain, id_scaling):
+                spike_stop_idx = spike_start_idx + convolution_resolution_len
+                objective.obj_normalized[overlapping_templates, spike_start_idx:spike_stop_idx] -= 2*pconv
+                if params.scale_amplitudes:
+                    pconv_scaled = pconv * spike_scaling
+                    objective.obj[overlapping_templates, spike_start_idx:spike_stop_idx] -= pconv_scaled
+
+            cls.enforce_refractory(spike_train, objective, params, template_meta)
+
+
+    @classmethod
+    def high_res_peak(cls, spike_time_indices, spike_unit_ids, objective, params, template_meta):
```

Comment:
`calculate_high_resolution_shift`? Iit is better to have functions as verbs

Reply:
legacy function name --> agree